### PR TITLE
impl(rest): improve error messaging from failed json to proto

### PR DESCRIPTION
**Pull request by scotthart**
Provide more detailed diagnostic when a json response from a REST call fails parsing into its intended protobuf Message.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10602)

### CODE REVIEW

File: google/cloud/internal/rest_stub_helpers.cc [hunk 1]
```diff
@@ -30,8 +30,13 @@ Status RestResponseToProto(google::protobuf::Message& destination,
   auto json_to_proto_status =
       google::protobuf::util::JsonStringToMessage(*json_response, &destination);
   if (!json_to_proto_status.ok()) {
-    return Status{
-        StatusCode::kInternal, std::string{json_to_proto_status.message()}, {}};
+    ErrorInfo err_info("Failure creating proto Message from Json",
+                       "google-cloud-cpp",
+                       {{"message_type", destination.GetTypeName()},
+                        {"json_string", *json_response}});
+    return Status{StatusCode::kInternal,
```
Comment (devbww) [inline google/cloud/internal/rest_stub_helpers.cc:37]: Should we try to preserve `json_to_proto_status.code()`?  [It looks like `absl::StatusCode` is already compatible with `google::cloud::StatusCode` via their mutual compatibility with the gRPC codes.]

Comment (coryan) [reply google/cloud/internal/rest_stub_helpers.cc:37]: I think I like @devbww idea of preserving the status code sounds sensible, maybe consider using `google::cloud::internal::ErrorInfoBuilder`

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10602)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/rest_stub_helpers.cc` line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10602#-NM5a7L3WTZhmM1KDDRK-r1-37:-NM5m8j07_BqPEXcjnvn:b-df9mu3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c8b0dffcdac5dc94f19e24a33924aed61d6aaa1a/google/cloud/internal/rest_stub_helpers.cc#L37)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think I like @devbww idea of preserving the status code sounds sensible, maybe consider using `google::cloud::internal::ErrorInfoBuilder`
</blockquote></details>

In our current version of proto the return type is a `protobuf::util::Status`.

**Review by devbww — COMMENTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10602)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan and @scotthart)

___
*[`google/cloud/internal/rest_stub_helpers.cc` line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10602#-NM5a7L3WTZhmM1KDDRK-r1-37:-NM5wvEu0aPZUB1shDCA:b-2x3hhu) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c8b0dffcdac5dc94f19e24a33924aed61d6aaa1a/google/cloud/internal/rest_stub_helpers.cc#L37)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

In our current version of proto the return type is a `protobuf::util::Status`.
</blockquote></details>

That's true, but even then `protobuf::util::StatusCode` had the same compatibility.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10602)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/rest_stub_helpers.cc` line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10602#-NM5a7L3WTZhmM1KDDRK-r1-37:-NM6B6W28lOMGKbUfCz_:b28usq3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c8b0dffcdac5dc94f19e24a33924aed61d6aaa1a/google/cloud/internal/rest_stub_helpers.cc#L37)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

That's true, but even then `protobuf::util::StatusCode` had the same compatibility.
</blockquote></details>

added static_cast

**Review by devbww — APPROVED** [review]

**Pull request merged**